### PR TITLE
[WIP] Fixed issue #15224: newQuestionAttributes deprecated : need a replacer

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: replace by getQuestionAttributes, because restricted by type
Dev: create QuestionAttribute::getQuestionAttributesPlugins to get the definition
Dev: add new definition in QuestionAttribute::model()->getQuestionAttributes
Dev: add new definition in Question::model()->getAdvancedSettingsWithValues

### CODE REVIEW

**Comment by Shnoulle** [comment]
Replace because previous one have 'type' new one get only by question type …

There still have incompatibility

````
    if(version_compare(Yii::app()->getConfig('versionnumber'),"4",">=")) {
        $scriptAttributes['settingsYesNo']['options']= array(
          'option'=> array(
            array('value'=>0, 'text'=> gT("No")),
            array('value'=>1, 'text'=> gT("Yes")),
          ),
        );
}
````
for example : an array in an array … 

Who made this XML definition : … 
    }

**Comment by olleharstedt** [comment]
You're fine with having one event being called in two places?

**Comment by olleharstedt** [comment]
I'm looking at the code and it's not easy to follow... Things being loaded everywhere. :(

**Comment by olleharstedt** [comment]
> Who made this XML definition : …

Me and Louis? There won't be a conflict, because Patrick added a feature that converts LS3 question theme to LS4 standard.

**Comment by olleharstedt** [comment]
This fix will cause the plugin event to be called twice in one request. getAdvancedSettingsWithValues calls getQuestionAttributes (includes event) and later getQuestionAttributesPlugins (event again).

File: application/models/Question.php [hunk 1]
```diff
@@ -292,14 +292,16 @@ public function getAdvancedSettingsWithValues($iQuestionID, $sQuestionType, $iSu
...
         // TODO: move getQuestionAttributesSettings() to QuestionAttribute model to avoid code duplication
         $aAttributeNames = QuestionAttribute::getQuestionAttributesSettings($sQuestionType);
 
         // If the question has a custom template, we first check if it provides custom attributes
-
         $oQuestion = Question::model()->find(array('condition'=>'qid=:qid', 'params'=>array(':qid'=>$iQuestionID)));
         $aAttributeNames = self::getQuestionTemplateAttributes($aAttributeNames, $aAttributeValues, $oQuestion);
-
+        // Add the questions attributes by plugins
+        $aAttributeNames = array_merge($aAttributeNames, QuestionAttribute::getQuestionAttributesPlugins($sQuestionType));
```
Comment (olleharstedt) [inline application/models/Question.php]: Why can't this line be removed if getQuestionAttributesPlugins is already called in getQuestionAttributes above?

Comment (Shnoulle) [reply application/models/Question.php]: One time for definition, one time for values …

Comment (olleharstedt) [reply application/models/Question.php]: But getQuestionAttributes already collect values here:
https://github.com/LimeSurvey/LimeSurvey/blob/master/application/models/QuestionAttribute.php#L240

**Comment by Shnoulle** [comment]
> > Who made this XML definition : …  
> Me and Louis? There won't be a conflict, because Patrick added a  feature that converts LS3 question theme to LS4 standard on upgrade (and  when imported).

Because we move to a simple XML
````
<options>
    <yes>Yes</yes>
    <no>No</no>
</options>
````
see https://github.com/tpartner/LimeSurvey-Net-Promoter-Score/blob/3fec6deceaac01839b3a5e90dbcea69e4100827c/Net-Promoter-Score/survey/questions/answer/listradio/config.xml#L72-L75

to
````
<options>
    <option>
        <value>yes</value>
        <text>Yes</text>
    </option>
    <option>
        <value>no</value>
        <text>No</text>
    </option>
</options>
````
https://github.com/LimeSurvey/LimeSurvey/blob/e865cdee3975491d76335d07844a8d024ce61f64/application/views/survey/questions/answer/gender/config.xml#L67-L76


Then in PHP move from
````
options = [
    'yes' => 'Yes,
    'no' => 'No,
];
````
to
````
options = [
    'option' => [
        'value' => 'yes',
        'text' => 'Yes'
    ]
    'option' => [
        'value' => 'no',
        'text' => 'No'
    ]
];
````

**Comment by Shnoulle** [comment]
> You're fine with having one event being called in two places?

Not really … but : i'm not fine with all question attribute related to question … only

**Comment by olleharstedt** [comment]
> Because we move to a simple XML [to complex]

OK, that I don't know. Check commit history?

**Comment by Shnoulle** [comment]
> I'm looking at the code and it's not easy to follow... Things being loaded everywhere. :(

You can not spoke for my code here … i add one function in 17 lines and 2 call of this function …

LimeSurvey new code for questionattribute are awfull … and i already say when it was merged it's awfull and complex …

**Comment by olleharstedt** [comment]
> You can not spoke for my code here … i add one function in 17 lines and 2 call of this function …

Yeah, I know.

> LimeSurvey new code for questionattribute are awfull … and i already say when it was merged it's awfull and complex …

Yeah. :(

**Comment by ptelu** [comment]
yeah this should be refactored, theres so much complexity and duplication going on its hard to navigate

**Comment by Shnoulle** [comment]
> yeah this should be refactored, theres so much complexity and duplication going on its hard to navigate

Currently : no question plugin can be used. This is clearly a big issue for user using CE edition. 

Again : i didn't choose the way to do the new QuestionAttribute. And this new way are : complex and hard to extend (see the issues with question theme : there are still some ).

Do you want i put all the mantis issue i do about new questionattribute system ?

**Comment by Shnoulle** [comment]
Oups https://github.com/LimeSurvey/LimeSurvey/pull/1370/commits/c67aeddb1bb4d6132ea1c310df6278ba654b3912

**Comment by olleharstedt** [comment]
Why did you close this?

**Comment by Shnoulle** [comment]
disablescriptwithxss is in this pull request …

**Comment by Shnoulle** [comment]
I mean this pull request : https://github.com/LimeSurvey/LimeSurvey/pull/1372
Are alreayd in this one

Replacer : https://github.com/LimeSurvey/LimeSurvey/pull/1376

**Comment by olleharstedt** [comment]
> Replacer : #1376

OK, thanks.

> I need it to git to "XSS for superadmin too" feature

Let's have a look on Monday.

**Pull request closed without merging**